File: lua/problem_54.lua
local M = {}

local function valid(board, r, c, d)
    for i = 1, 9 do
        if board[r][i] == d or board[i][c] == d then return false end
    end
    local br = math.floor((r-1)/3)*3+1
    local bc = math.floor((c-1)/3)*3+1
    for i = 0, 2 do for j = 0, 2 do
        if board[br+i][bc+j] == d then return false end
    end end
    return true
end

local function solve(board)
    for r = 1, 9 do for c = 1, 9 do
        if board[r][c] == 0 then
            for d = 1, 9 do
                if valid(board, r, c, d) then
                    board[r][c] = d
                    if solve(board) then return true end
                    board[r][c] = 0
                end
            end
            return false
        end
    end end
    return true
end

function M.solve_sudoku(board) solve(board); return board end

function M.problem54()
    print("Problem 54:")
    local board = {
        {5,3,0,0,7,0,0,0,0},{6,0,0,1,9,5,0,0,0},{0,9,8,0,0,0,0,6,0},
        {8,0,0,0,6,0,0,0,3},{4,0,0,8,0,3,0,0,1},{7,0,0,0,2,0,0,0,6},
        {0,6,0,0,0,0,2,8,0},{0,0,0,4,1,9,0,0,5},{0,0,0,0,8,0,0,7,9},
    }
    M.solve_sudoku(board)
    for r = 1, 9 do for c = 1, 9 do assert(board[r][c] ~= 0) end end
    print("All Tests Passed")
end

return M

File: c/problem_54.c
#include "problem_54.h"
#include <stdio.h>
#include <assert.h>

static int valid54(int b[9][9], int r, int c, int d) {
    for (int i=0;i<9;i++) if(b[r][i]==d||b[i][c]==d) return 0;
    int br=(r/3)*3, bc=(c/3)*3;
    for (int i=0;i<3;i++) for(int j=0;j<3;j++) if(b[br+i][bc+j]==d) return 0;
    return 1;
}

int solve_sudoku(int board[9][9]) {
    for (int r=0;r<9;r++) for(int c=0;c<9;c++) {
        if (board[r][c]!=0) continue;
        for (int d=1;d<=9;d++) {
            if (valid54(board,r,c,d)) {
                board[r][c]=d;
                if (solve_sudoku(board)) return 1;
                board[r][c]=0;
            }
        }
        return 0;
    }
    return 1;
}

void problem54(void) {
    printf("Problem 54:\n");
    int board[9][9] = {
        {5,3,0,0,7,0,0,0,0},{6,0,0,1,9,5,0,0,0},{0,9,8,0,0,0,0,6,0},
        {8,0,0,0,6,0,0,0,3},{4,0,0,8,0,3,0,0,1},{7,0,0,0,2,0,0,0,6},
        {0,6,0,0,0,0,2,8,0},{0,0,0,4,1,9,0,0,5},{0,0,0,0,8,0,0,7,9},
    };
    assert(solve_sudoku(board));
    for(int r=0;r<9;r++) for(int c=0;c<9;c++) assert(board[r][c]!=0);
    printf("All Tests Passed\n");
}
